sprint_11

igorysssha opened the pull request:
Сдача 11 спринта

igorysssha committed:
Добавил в singleViewController функцию scrollViewDidZoom, которая центрирует картинку после зума

igorysssha committed:
20.09

igorysssha committed:
Сдача 11 sprint

igorysssha committed:
добавил kingfisher и swiftkeycjain

romanenko-ivan commented on ImageFeedProject/SplashViewController/SplashViewController.swift:
```diff
@@ -33,38 +74,101 @@ class SplashViewController: UIViewController {
+        isFetchingProfile = true
+        DispatchQueue.main.async {
+            UIBlockingProgressHUD.show()
+            self.profileService.fetchProfile(token) { [weak self] result  in
```
Cильно захвачен `self`

romanenko-ivan commented on ImageFeedProject/Profile/ProfileViewController.swift:
```diff
@@ -1,97 +1,157 @@
+        ])
+        
+        // label с именем and constraints
+        let nameLabel = UILabel()
```
Могу предложить кастомизировать свойства прямо при объявлении через замыкание:

```
let topLineView: UIView = {
    let lineView = UIView()
    lineView.layer.borderWidth = 1.0
    lineView.layer.borderColor = UIColor.lightGray.cgColor
    return lineView
}()

```
Таким образом вся настройка компонента будет в одном месте, а в методе, например `configureView()`, выполнить настройку всего лейаута(`addSubview()` и `NSLayoutConstraint.activate([])`)

romanenko-ivan replied:
Больше подробностей о том, как можно упорядочить код верстки [здесь](https://medium.com/@rckim77/better-programmatic-components-using-closures-50d1086f8b79)

romanenko-ivan commented on ImageFeedProject/Auth/WebViewViewController.swift:
```diff
@@ -27,53 +28,36 @@ final class WebViewViewController: UIViewController {
-        URLQueryItem(name: "redirect_uri", value: RedirectURI),
-        URLQueryItem(name: "response_type", value: "code"),
-        URLQueryItem(name: "scope", value: AccessScope)
+            URLQueryItem(name: "client_id", value: AccessKey),
```
Имена функций и переменных принято объявлять в нижнем `camelCase`, а название типов классов, структур и перечислений в верхнем `CamelCase`. `AccessKey` и другие константы - следует переименовать. Хорошей практикой будет занести все константы в энам, например:

```
enum Constants {
  static let accessKey = ""
  ...
}
```

romanenko-ivan commented on ImageFeedProject/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+}
+
+struct Profile {
```
Давай разделять код на файлы по следующей логике: 1 файл - 1 класс/структура/модель и т.д.

romanenko-ivan commented on ImageFeedProject/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,113 @@
+    let followedByUser: Bool
+    let downloads: Int
+    
+    enum CodingKeys: String, CodingKey {
```
Предлагаю использовать настройку декодера вместо использования `CodingKeys`

https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable — настройки JSONDecoder

romanenko-ivan commented on ImageFeedProject/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    private func makeProfileRequest(withToken token: String) -> URLRequest? {
+        guard let url = URL(string: "https://api.unsplash.com/me") else {
```
Давай вынесем в константы

romanenko-ivan commented on ImageFeedProject/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+    
+    private func makeProfileRequest(username: String, token: String) -> URLRequest? {
+        let urlString = "https://api.unsplash.com/users/\(username)"
```
Адрес лучше получать из констант

romanenko-ivan commented on ImageFeedProject/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -28,31 +36,51 @@ final class OAuth2Service {
+                    self.authToken = authToken
+                    completion(.success(authToken))
+                case .failure(let error):
+                    print("[OAuth2Service]: Ошибка получения токена - \(error.localizedDescription)")
```
Можно обратиться к функции, чтобы получить ее название, здесь больше подробностей как можно улучшить логирование через `print` https://www.appypie.com/print-swift-how-to

romanenko-ivan commented on ImageFeedProject/TabBarController/TabBarController.swift:
```diff
@@ -0,0 +1,25 @@
+                                                        image:UIImage(named: "tab_profile_active"),
+                                                        selectedImage: nil)
+        
+        self.viewControllers = [imagesListViewController, profileViewController]
```
Явный `self` можно опустить

romanenko-ivan commented on ImageFeedProject/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,96 @@
+
+final class ProfileImageService: ProfileImageServiceProtocol {
+    static let didChangeNotification = Notification.Name(rawValue: "ProfileImageProviderDidChange")
+    private(set) var avatarURL: String?
```
Стоит организовывать код следующим образом:

  - Публичные переменные
  - IBOutlet
  - Приватные переменные
  - Публичные методы(сначала initы, overrideы, и потом остальные)
  - Приватные методы
  - IBAction

Почитать подробнее про организация кода:
https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization

romanenko-ivan commented on ImageFeedProject/Profile/ProfileViewController.swift:
```diff
@@ -1,97 +1,157 @@
+                self.updateAvatar()
+            }
+        updateAvatar()
+        //print("Small Image URL: \(String(describing: ProfileImageService.shared.avatarURL))")
```
Закомментированный код лучше убрать

igorysssha committed:
Готово, изменил все критические замечания из списка
В методах SplashViewController.fetchProfile(на строке 108) и ProfileViewController.updateUIWithProfile сильно захвачен self.

убрал implicit unwrap в:

при объявлении DefaultBaseURL;
методе makeHTTPRequest;
при создании свойства logoutButton в методе setupLabels;
методе ImagesListViewController.prepare в строке let indexPath = sender as! IndexPath;
при указании типа свойству image в SingleImageViewController;
методе didTapShareButton)
А также некоторые рекомендации(обязательно доделаю все)

igorysssha committed:
и тут

romanenko-ivan commented on ImageFeedProject/ImagesList/ImagesListViewController.swift:
```diff
@@ -31,15 +31,21 @@ class ImagesListViewController: UIViewController {
+                let viewController = segue.destination as? SingleImageViewController,
+                let indexPath = sender as? IndexPath
+            else {
+                assertionFailure("Не удалось выполнить переход к SingleImageViewController")
```
Использование выражений не приветствуется внутри основного кода приложения, лучше только в тестах:
https://stevenpcurtis.medium.com/precondition-assert-fatal-error-or-guard-in-your-swift-code-5f9297658be0
